tools/clang/lib/SPIRV/SpirvEmitter.cpp

```diff
@@ -11937,15 +11937,6 @@ bool SpirvEmitter::spirvToolsOptimize(std::vector<uint32_t> *mod,
     // Add performance passes.
     optimizer.RegisterPerformancePasses();
 
-    // Add flattening of resources if needed.
```

Comment:
Is it ok that we do not apply this when `-O3` option is given?

Reply:
it _will_ be run with `-O3` because `-O3` requires legalization.

Reply:
Do you mean `-O3` will set [`needsLegalization`](https://github.com/microsoft/DirectXShaderCompiler/blob/master/tools/clang/lib/SPIRV/SpirvEmitter.cpp#L591) true?

Reply:
no. Sorry, maybe I didn't explain it properly.

Note that `needsLegalization` is always true if the SPIR-V backend finds a global struct of resources. See [here](https://github.com/microsoft/DirectXShaderCompiler/blob/f1f60648df03f20b1281bfbfbd356c4f6d9f3b48/tools/clang/lib/SPIRV/SpirvEmitter.cpp#L594).

What I mean is: There is no way to run optimizations without legalization. There are 3 code-paths:
1) don't legalize and don't optimize
2) legalize, but don't optimize
3) legalize and optimize

In the past, we were running "descriptor scalar replacement" during optimization. Now we want to do it during legalization.

For  path 1: didn't run the pass before my change, and won't run it now
For  path 2: didn't run the pass before my change, but will run it now --> which will result in the behavior that the user wants in issue #3100 
For  path 3: Ran the pass before my change, and will run it with my change too.

Does it make sense?

Reply:
Gotcha.